Added Sidebar CreateMenu Drawer

lucas-neuhaus-dev opened the pull request:
@marcaaron 

### Details
Created a CreateMenu Modal. it should be displayed when the floating action button is pressed and should animate from the bottom of the sidebar. 

The modal closes either when the user press one of its items, or press anywhere outside of it on the screen.

When the user presses one of the modal's items, it closes the modal, and activates the chat switcher (mirrors the action when pressing on the search icon on the top of the sidebar).

#### Problem found and how I fixed

**_This problem was fixed with other commits on the `master branch`, the trick below is not needed anymore and was fixed_**

As I mentioned on the [issue page](https://github.com/Expensify/Expensify.cash/issues/1240), there was a problem when calling the `ChatSwitcher.show()` (when pressing one of the modal's items). From my debugging, for the desired result to occur, the app also needs to call a couple of functions declared on the `src/pages/home/sidebar/ChatSwitcherView.js`. It needed to call both `updateSearch()` and `textInput.focus()`.

What I did was: I added (on the file cited above) a check for change on the `isChatSwitcherActive`, checking for a change from false to true. This check happens on `componentDidUpdate()`. If it is true, `updateSearch()` and `textInput.focus()` are called. 

#### Divergences from the initial proposal

I said first that the `CreateMenu` component would be called on the `HomePage.js` file. Decided to change it and call it on the `SidebarView.js` file. It is cleaner this way in my opinion.

#### Not really a problem but I changed

When I was debugging the first problem, I noticed the the `isChatSwitcherActive` prop was being passed down all the way from the `HomePage`. I don't believe these variables stored in Onyx should be passed down this way, I think it is much cleaner to always add it to the `withOnyx` HOF to access it. 

When I got to the `src/pages/home/sidebar/ChatSwitcherView.js` file I actually thought other files could be passing it down wrong and this could be the cause of the problem. After going all the way up to `HomePage` I realized it was correct.

What I mean is, when you see the prop is being accessed via `withOnyx()`, you know it could not be the issue, and you are not sure otherwise.

I changed the `isChatSwitcherActive` be accessed with `withOnyx` on the files I worked with, but did not looked much further to see if it happens in other locations.

If the way it was is the desired way, not a problem. I can undo the changes I made in an instant.


### Fixed Issues
Fixes #1240 

### Tests
1. Click on the floating action button
2. Verify a modal opens with a menu animating from the bottom. The menu must contain 2 items with icon and text
3. When clicking any of the menu's items, it should activate the chat switcher (action should be the same as when clicking the 'search' icon on the top of the sidebar), and the modal should close
4. If clicking anywhere on the screen outside the menu, the modal should close

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->


#### Web
![web](https://i.ibb.co/YQvLVLW/web.gif)

#### Mobile Web
![mobile web](https://i.ibb.co/FDpm4qm/mobile-web.gif)

#### Desktop
![desktop](https://i.ibb.co/GRPzqbD/desktop.gif)

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
![android](https://i.ibb.co/t3d6nsM/android.jpg)

lucas-neuhaus-dev committed:
Added Sidebar CreateMenu Drawer

marcaaron commented:
@lucas-neuhaus-dev you've got some conflicts I can take a look at this once they are resolved thanks!!

marcaaron commented on src/pages/home/sidebar/ChatSwitcherView.js:
```diff
@@ -106,6 +108,13 @@ class ChatSwitcherView extends React.Component {
+        if (!prevProps.isChatSwitcherActive && this.props.isChatSwitcherActive) {
+            this.updateSearch(this.state.search);
+            this.textInput.focus();
+        }
```
I think  merging master will help with this since we are conditionally rendering the view now instead of toggling the active state of the "chat switcher".

lucas-neuhaus-dev replied:
Noow I understand what you mean 🤦‍♂️

This is really not necessary anymore. Just calling ChatSwitcher.show() does the whole trick now.

lucas-neuhaus-dev replied:
I believe there is still something to fix on this conditional render @marcaaron.
When the ChatSwitcher **is active**, and `ChatSwitcher.show()` is called, the text input does not focus again.

![bug](https://i.ibb.co/yNLZ0bV/croped-bug.gif)

lucas-neuhaus-dev committed:
Merged master branch and fixed conflicts

lucas-neuhaus-dev commented:
@marcaaron Merged master branch and:

1. Fixed the conflicts;
2. Created new animated gifs with tests on the new UI on the description;
3. Added the **Tests** part of the description (I forgot to write it before);
4. I had described a bug on the description when activating the chat switcher, removed it since it has been fixed.

Julesssss commented on src/pages/home/HomePage.js:
```diff
@@ -61,19 +62,33 @@ class App extends React.Component {
+        // while including mutual callbacks first
+        this.menuItemData = [
+            {icon: ChatBubbleIcon, text: 'New Chat', onPress: () => {}},
+            {icon: UsersIcon, text: 'New Icon', onPress: () => {}},
```
This should say 'New Group'

lucas-neuhaus-dev replied:
Pushed an amend fixing it.

Julesssss commented on src/pages/home/HomePage.js:
```diff
@@ -61,19 +62,33 @@ class App extends React.Component {
+            !props.isSidebarShown ? -variables.sideBarWidth : 0,
         );
+
+        // This format allows to set individual callbacks to each item
+        // while including mutual callbacks first
```
Just so I understand... each item's callback will eventually be set here right? but because they currently have the same behaviour they are empty for now?

lucas-neuhaus-dev replied:
I was not sure where the `menuItemData` array was meant to be declared, that's why I ended up writing it on the `HomePage` constructor. 
The [issue page](https://github.com/Expensify/Expensify.cash/issues/1240) just stated that "all menu selections should call `ChatSwitcher.show()`". I did the `Array.map()` trick so we have a template when we eventually need to call different actions for different items, but might still have a callback in common.

marcaaron replied:
This is just a temporary step. We will add two new views New Chat and New Group that will open modally in the future. But yes we should have two separate callbacks..

Julesssss reviewed (dismissed):
Looking good.

One thing I noticed is that the input field highlight only works once: If you open the menu and select 'New Chat' then the search component is displayed and the input field is highlighted. But if you open the menu a second time and select 'New Chat' the input field remains unfocused. 

_Edit: Just saw that you've already discussed this above_

Julesssss commented:
One other minor thing is that the menu item text style looks a little off (too dark compared to icon colour?), but I could be wrong about this. 

@Expensify/design could you please confirm whether the style is correct? Thanks!

![Screenshot 2021-01-18 at 15 08 36](https://user-images.githubusercontent.com/10736861/104935919-bff55600-59a3-11eb-94ba-8fc28c095f11.png)

lucas-neuhaus-dev commented:
@Julesssss 

> One thing I noticed is that the input field highlight only works once: If you open the menu and select 'New Chat' then the search component is displayed and the input field is highlighted. But if you open the menu a second time and select 'New Chat' the input field remains unfocused.

You are right. I noticed it as well and even mentioned it on the @marcaaron open issue above. I am going to take a look at it now see if I can come up with a clean fix.

> One other minor thing is that the menu item text style looks a little off (too dark compared to icon colour?), but I could be wrong about this.

I tried to replicate the screenshot given on the [issue page](https://github.com/Expensify/Expensify.cash/issues/1240), with the existing colors declared on `styles`. But I agree with you I could have done it better. 
Especially on hover, I think the gray should be lighter, I didn't use a lighter one because I would need to create a new color, and I don't know how strict the color palette is.
I would appreciate clearer instructions from the design team, but I will try some stuff out.

marcaaron commented:
>I am going to take a look at it now see if I can come up with a clean fix

There's no need. Eventually we will open separate views for these two menu items. Opening the chat switcher is only a temporary step so we don't need to spend time on re-focusing the text input.

michelle-thompson commented:
@Julesssss yup, this definitely looks like it's black. I'd suggest `UI/Heading` for these!

lucas-neuhaus-dev committed:
changed menu item text color

lucas-neuhaus-dev commented:
@michelle-thompson 
Changed the color to `themeColors.heading`. Here is the preview:
![new text color](https://i.ibb.co/sPNWNyV/new-item-color.png)

I will update all the images/gifs in the description once we get the desired style agreed.

michelle-thompson commented:
@lucas-neuhaus-dev Can you double check which type this is?

lucas-neuhaus-dev commented:
@michelle-thompson Of course. One important note, I really thought I was using the lightest gray on `styles` for the hover effect, but I was not. The next preview is using the lightest gray (not yet pushed on the PR).
Preview:
![preview with hover](https://i.ibb.co/Jtw5KtH/preview-with-hover-bg.png)

Colors: 
Icon: `themeColors.icon` - Which corresponds to `colors.gray3` - `#C6C9CA`
Text: `themeColors.heading` - Which corresponds to `colors.charcoal` - `#37444C`
Hover: `themeColors.buttonHoveredBG` _I added it on `src/styles/themes/defaults.js`_ - Which corresponds to `colors.gray1` - `#FAFAFA`

shawnborton commented:
The spacing and padding is still incorrect. 

Here is the anatomy of a menu row:
![image](https://user-images.githubusercontent.com/2319350/104950464-8efe2c80-5975-11eb-8b16-cdd98cd32be4.png)

The inner contents should be 40 tall
The padding around the inner contents is 20 horizontal, and 12 vertical
The menu icon should be wrapped in a 40x40 area
The menu icon itself (inside of the 40x40 area) should be 24x24
The gap between the menu icon and the label should be 12
The menu icon color should be `themeColors.icon`
The label color should be `themeColors.heading`

There should also be 12 on top/bottom of the group of menu icons:
![image](https://user-images.githubusercontent.com/2319350/104950612-d389c800-5975-11eb-84d0-35b0bd5579f4.png)


Then at the bottom, you need to include a SafeAreaPadding or SafeAreaMargin (I forget what it is) for the home bar on iPhone X

shawnborton commented:
I like using the lightest gray for the hover background color 👍

lucas-neuhaus-dev commented:
I am pushing a commit that gets **almost** all the sizes mentioned by @shawnborton right.
It is just almost because of the way the `Modal` component is written. And because I believe for the final menu there will be a new Modal component, I did not change the current one.

The current modal (`src/components/Modal.js`) component has this View as a child:
```
<View
    style={{
        ...styles.defaultModalContainer,
        paddingBottom,
        ...modalContainerStyle,

        // This padding is based on the insets and could not neatly be
        // returned by getModalStyles to avoid passing this inline.
        paddingTop: needsSafeAreaPadding ? paddingTop : 20,    // <--- This line
    }}
>
    {props.children}
</View>
```
This makes it impossible to change the container's top padding without rewriting it (since the other styles are passed before).

I will wait for further instructions on this.

**New Preview:**
![preview](https://i.ibb.co/7gYyKGj/new-hover.png)

**Side note:** I don't want this to sound as a criticism, just as a suggestion (and I should have asked for it before). I think sharing a Figma file of the component when a freelancer is working one it, or a detailed image with the dimensions and colors as posted now, would go a long way in optimizing the back-and-forth between the freelancer and the development team.

lucas-neuhaus-dev committed:
fixed create menu style

marcaaron commented:
>This makes it impossible to change the container's top padding without rewriting it (since the other styles are passed before).

If I am understand correctly, we want to apply a `paddingTop` in the `modalContainerStyle` but the hardcoded `paddingTop` in this line prevents it? 

https://github.com/Expensify/Expensify.cash/blob/d8867efd8456886aa1293cdac199f0d4a74890e0/src/components/Modal.js#L67-L69

If that's correct, then we can just move this to before we spread in the `modalContainerStyle` so we can pass in whatever values we like. Let me know if I am misunderstanding.

lucas-neuhaus-dev commented:
@marcaaron You are correct! I was just wanting to make sure before changing the component.
Moved the code up and fixed the issue.

lucas-neuhaus-dev commented:
Images and animations updated on the PR description.

shawnborton commented on src/styles/styles.js:
```diff
@@ -406,6 +406,27 @@ const styles = {
+
+    sidebarCreateMenuText: {
+        fontFamily: fontFamily.GTA_BOLD,
+        fontSize: variables.fontSizeLabel,
```
Can we actually make this fontSizeNormal? So it should be 15.

lucas-neuhaus-dev replied:
Done!

shawnborton commented:
Nice, this is looking great! Left one small comment for you regarding font size on the menu labels. 

I tested this out on web for now, and my initial feedback is:
- can we use some kind of `cursor: pointer` for the menu options?
- there are times when launching the menu feels very laggy. Basically I just kept interacting with it a bunch, and there were times where the animation felt very slow and choppy. Not sure what's going on performance-wise there, but worth looking into.

lucas-neuhaus-dev commented:
@shawnborton 

> can we use some kind of cursor: pointer for the menu options?

The cursor should already be showing the as a pointer. It is the default on `Pressable` from `react-native-web` version `0.14.1` and forward. It works correctly for me on web and desktop. Could you please double-check and see if it is really not showing right for you?

> there are times when launching the menu feels very laggy.
 
We can try to turn `useNativeDriver` and `useNativeDriverForBackdrop` on the Modal to `true` _([react-native-modal API](https://github.com/react-native-modal/react-native-modal)_). I am going to take a look at the code again, see if I can think of something. For now, that is the only thing that crosses my mind.

I am going to push a commit with these props.

lucas-neuhaus-dev committed:
Modal uses native driver

marcaaron commented on src/styles/styles.js:
```diff
@@ -406,6 +406,27 @@ const styles = {
+        alignItems: 'center',
+    },
+
+    sidebarCreateMenuText: {
```
Can probably lose the `sidebar` on all of these styles.

lucas-neuhaus-dev replied:
Done!

marcaaron commented on src/pages/home/sidebar/SidebarView.js:
```diff
@@ -14,39 +17,64 @@ const propTypes = {
+        <CreateMenu
+            onClose={props.toggleCreateMenu}
+            isVisible={props.isCreateMenuActive}
+            menuItemData={props.menuItemData}
```
This is maybe deviating a bit from the original proposal, but thinking about this some more and we are drilling these props down a bit unnecessarily. Here's an alternative idea...

1. Declare the menu items inside `CreateMenu`
2. Add an `onItemSelected` callback
3. Pass a function down to `CreateMenu` instead of having to declare the items in `HomePage`.

Feel free to push back on that as we can always improve this later.

lucas-neuhaus-dev replied:
I have to say I wasn't liking the `menuItemData` array on the `HomePage` as well.
I guess it can make it harder to get each item's callback function when they have one, but it's definitely worth trying out to see.

I put a more declarative name for the function on the `HomePage`: `onCreateMenuItemSelected`, left it as `onItemSelected` inside the `CreateMenu` component.

I wrapped `CreateMenu` with `React.memo()` now, since I don't want the `menuItemData` array to be reconstructed as often.

marcaaron replied:
Nice it looks good!

marcaaron reviewed (changes requested):
Looking great so far. I did not experience any animation issues and cursor pointer works fine on web. I did have some other comments.

lucas-neuhaus-dev commented on src/pages/home/sidebar/SidebarView.js:
```diff
@@ -14,39 +17,64 @@ const propTypes = {
-export default SidebarView;
+export default withOnyx(
+    {
+        isChatSwitcherActive: {
```
No problem. I explained why I did that on the PR description, on the **_Not really a problem but I changed it_** section. For me kind of make sense to explicitly say that we are getting the prop from the store on each file _(single source of truth kind of deal)_. But that was just a suggestion.

I am going to the gym, but I will finish these reviews right when I'm back :thumbsup:.

marcaaron replied:
Ah missed that. My instinct says to only add another subscriber if we must, but we haven't really settled on a best practice yet. :)

lucas-neuhaus-dev replied:
If subscribing to the component does have a performance cost, it makes sense.

lucas-neuhaus-dev commented on src/components/CreateMenu.js:
```diff
@@ -0,0 +1,61 @@
+                ])}
+            >
+                <View style={styles.sidebarCreateMenuIcon}>
+                    <item.icon width={24} height={24} />
```
`<item.icon />` really is ugly. I agree.

lucas-neuhaus-dev committed:
Declared menuItemData on the CreateMenu component

marcaaron commented on src/components/Modal.js:
```diff
@@ -52,6 +53,8 @@ const Modal = (props) => {
             animationIn={animationIn}
             animationOut={animationOut}
+            useNativeDriver
+            useNativeDriverForBackdrop
```
I'm not sure but these are possibly causing some flickering animations on iOS. I commented out these lines and things smoothed out again.

https://github.com/react-native-modal/react-native-modal#the-modal-flashes-in-a-weird-way-when-animating

lucas-neuhaus-dev replied:
Went on the `react-native-modal` Github page, and found a couple ([[1](https://github.com/react-native-modal/react-native-modal/issues/268)] [[2](https://github.com/maxs15/react-native-modalbox/issues/265)]) of issues on this topic.

There seems to have some workarounds if more people complain of performance issues, and we want to test out these props again.

Removed it for now.

marcaaron reviewed (dismissed):
Everything looks good to me. I think we should just remove the lines about `useNativeDriver`.

lucas-neuhaus-dev commented:
Removed the `useNativeDriver` and `useNativeDriverForBackdrop` props from the Modal.

marcaaron commented:
I'm going to merge this so we can keep progressing. If there are any last comments to add this week please let us know @Julesssss @shawnborton

Julesssss reviewed (commented):
Latest changes looks good to me 👍
